### PR TITLE
Icravideo

### PR DESCRIPTION
**Pull request by mehtank**

### CODE REVIEW

File: python/icra_double.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
+import random
```
Comment (cathywu) [inline python/icra_double.py]: Should we get rid of icra_single and double? are they still useful outside of icra.py?

File: python/loopsim.py [hunk 2]
```diff
@@ -69,22 +70,26 @@ def _mkdirs(self, name):
...
-        self.sumoProcess = subprocess.Popen([
+        sumoBinary = checkBinary(sumo)
+        sumoProcessArgs = [
                 sumoBinary, 
                 "--step-length", repr(self.simStepLength),
                 "--no-step-log",
                 "-c", self.cfgfn,
-                "--remote-port", str(self.port)], 
+                "--remote-port", str(self.port)]
+        if sublane:
```
Comment (cathywu) [inline python/loopsim.py:90]: What is the sublane model used for?

File: python/loopsim.py [hunk 3]
```diff
@@ -158,13 +190,18 @@ def _run(self, simSteps):
...
                 self.allCars.append(car)
             self.allCars.sort(key=lambda x: x["x"])
 
             for (idx, car) in enumerate(self.allCars):
+                self._setCarColor(car, speedRange)
                 carFn = self.carFns[car["type"]]
                 if carFn is not None:
                     carFn((idx, car), self, step)
+            if sumo == "sumo-gui":
+                traci.gui.screenshot("View #0", "%s/%08d.png" % (vid_path, step))
```
Comment (cathywu) [inline python/loopsim.py:206]: What does this do? (Can you add a comment?)

File: python/loopsim.py [hunk 4]
```diff
@@ -158,13 +190,18 @@ def _run(self, simSteps):
                 car["lane"] = traci.vehicle.getLaneIndex(v)
                 car["x"] = self._getX(car["edge"], position)
                 car["v"] = traci.vehicle.getSpeed(v)
+                car["maxv"] = traci.vehicle.getMaxSpeed(v)
+                car["f"] = traci.vehicle.getSpeedFactor(v)
```
Comment (cathywu) [inline python/loopsim.py:194]: This doesn't change throughout the simulation, so can we retrieve and store this once?

File: python/makecirc.py [hunk 5]
```diff
@@ -15,31 +18,45 @@ def makexml(name, nsl):
 def printxml(t, fn):
     etree.ElementTree(t).write(fn, pretty_print=True, encoding='UTF-8', xml_declaration=True) 
 
-def makenet(base, length, lanes, speedLimit=defaults.SPEED_LIMIT, path=""):
-    import subprocess
-    import sys
+def makenet(base, length, lanes, 
+        speedLimit=defaults.SPEED_LIMIT,
```
Comment (cathywu) [inline python/makecirc.py:22]: nit: why are these now on new lines?

File: python/makecirc.py [hunk 6]
```diff
@@ -145,8 +168,12 @@ def outputs(name, prefix="data/"):
     else:
         roufn=False
 
+    gui = E("viewsettings")
+    gui.append(E("scheme", name="real world"))
```
Comment (cathywu) [inline python/makecirc.py:172]: Beautiful.

File: python/parsexml.py [hunk 7]
```diff
@@ -57,24 +58,48 @@ def parsexml(fn, edgestarts, xmax, vdefault=0):
             pass
 
         for lid, thislane in lanedata.iteritems():
-            f = interp([x["x"] for x in thislane], [x["v"] for x in thislane], xmax, vdefault)
+            vf = interp([x["x"] for x in thislane], [x["v"] for x in thislane], xmax, vdefault)
```
Comment (cathywu) [inline python/parsexml.py:62]: What does f mean? Frame? Can you add a comment?

File: python/plots.py [hunk 8]
```diff
@@ -61,71 +61,128 @@ def pcolor(title, (xrng, xlabel),
...
+    '''
     # Add colorbar
-    fig.subplots_adjust(right=0.8)
-    cbar_ax = fig.add_axes([0.84, 0.35, 0.02, 0.55])
-    cbar_ax2 = fig.add_axes([0.84, 0.1, 0.02, 0.2])
-
-    ticks = linspace(smin, smax, 6)
+    fig.subplots_adjust(right=0.85)
+    if numlanes == 2:
+        cbm = 0.52
```
Comment (cathywu) [inline python/plots.py:172]: What are `cbm`, `cbl`?

**Comment by mehtank** [comment]
On Mon, Sep 19, 2016 at 05:36:25PM -0700, Cathy Wu wrote:

> cathywu commented on this pull request.
> 
> > @@ -0,0 +1,74 @@
> > +import random
> 
> Should we get rid of icra_single and double? are they still useful outside of icra.py?

Gone.  Should keep in icrasims branch though, because they're used to
make the paper graphics.

> > ```
> >              sumoBinary, 
> >              "--step-length", repr(self.simStepLength),
> >              "--no-step-log",
> >              "-c", self.cfgfn,
> > ```
> > -                "--remote-port", str(self.port)], 
> > -                "--remote-port", str(self.port)]
> > -        if sublane:
> 
> What is the sublane model used for?

It's meant to allow for smooth instead of instantaneous lane changes.
Turns out cars often just get stuck between lanes for some reason.
Probably need to set some parameters somewhere?

> > ```
> >              carFn = self.carFns[car["type"]]
> >              if carFn is not None:
> >                  carFn((idx, car), self, step)
> > ```
> > -            if sumo == "sumo-gui":
> > -                traci.gui.screenshot("View #0", "%s/%08d.png" % (vid_path, step))
> 
> What does this do? (Can you add a comment?)

Added.

> > @@ -158,13 +190,18 @@ def _run(self, simSteps):
> >                  car["lane"] = traci.vehicle.getLaneIndex(v)
> >                  car["x"] = self._getX(car["edge"], position)
> >                  car["v"] = traci.vehicle.getSpeed(v)
> > -                car["maxv"] = traci.vehicle.getMaxSpeed(v)
> > -                car["f"] = traci.vehicle.getSpeedFactor(v)
> 
> This doesn't change throughout the simulation, so can we retrieve and store this once?

Umm.  Yeah, we could.  There are probably other speedup tweaks that
would have a bigger effect though, like the sort() immediately
following.  I'm not sure it's worth doing yet.  

> > @@ -15,31 +18,45 @@ def makexml(name, nsl):
> >  def printxml(t, fn):
> >      etree.ElementTree(t).write(fn, pretty_print=True, encoding='UTF-8', xml_declaration=True) 
> 
> -def makenet(base, length, lanes, speedLimit=defaults.SPEED_LIMIT, path=""):
> -    import subprocess
> -    import sys
>   +def makenet(base, length, lanes, 
> -        speedLimit=defaults.SPEED_LIMIT, 
> 
> nit: why are these now on new lines?

It looked nicer to me that way.

> > @@ -145,8 +168,12 @@ def outputs(name, prefix="data/"):
> >      else:
> >          roufn=False
> -    gui = E("viewsettings")
> -    gui.append(E("scheme", name="real world"))
> 
> Beautiful.
> 
> > @@ -57,24 +58,48 @@ def parsexml(fn, edgestarts, xmax, vdefault=0):
> >              pass
> 
> ```
>      for lid, thislane in lanedata.iteritems():
> ```
> -            f = interp([x["x"] for x in thislane], [x["v"] for x in thislane], xmax, vdefault)
> -            vf = interp([x["x"] for x in thislane], [x["v"] for x in thislane], xmax, vdefault)
> 
> What does f mean? Frame? Can you add a comment?

f for function.  Comment added.

> > ```
> >  # Add colorbar
> > ```
> > -    fig.subplots_adjust(right=0.8)
> > -    cbar_ax = fig.add_axes([0.84, 0.35, 0.02, 0.55])
> > 
> > ## \-    cbar_ax2 = fig.add_axes([0.84, 0.1, 0.02, 0.2])
> > -    ticks = linspace(smin, smax, 6)
> > -    fig.subplots_adjust(right=0.85)
> > -    if numlanes == 2:
> > -        cbm = 0.52
> 
> What are `cbm`, `cbl`?

Random numbers to make the colorbar plot positioned in the right place.
Added comment.

**Pull request merged**